doc/design/eventdriven-callbacks.md

```diff
@@ -0,0 +1,215 @@
+    - [robust] Updating a callback from within the same callback 
+      ```cpp
+      void sampleReceived2(iox::popo::UntypedSubscriber & subscriber) {}
+
+      void sampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+        myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, sampleReceived2);
+      }
+
+      myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, sampleReceived);
+      ```
+ - concurrently: detach a callback at any time from anywhere. This means in particular
+    - Detaching a callback from within a callback.
+    - [robust] When the detach call returns we guarantee that the callback is never called
+      again. E.g. blocks till the callback is removed, if the callback is concurrently 
+      running it will block until the callback is finished and removed.
+      Exception: If a callback detaches itself it blocks until the callback is removed 
+                 and not until the callback is finished.
+    - Calling detach means that after that call the callback is no longer attached
+      even when it was not attached in the first place. Therefore it will always succeed.
+
+ - When the ReactAl goes out of scope it should detach itself from every class 
+     to which it was attached via a provided callback (like in the WaitSet).
+
+ - When the class which is attached to the ReactAl goes out of scope it should 
+    detach itself from the ReactAl via a provided callback (like in the WaitSet).
+
+### Usage Code Example
+```cpp
+ReactAl myCallbackReactal;
+iox::popo::UntypedSubscriber mySubscriber;
+
+void sampleReceived(iox::popo::UntypedSubscriber & subscriber) {
+  subscriber.take().and_then([&](auto & sample){
+    std::cout << "received " << std::hex << sample->payload() << std::endl;
+  });
+}
+
+void waitForSubscription(iox::popo::UntypedSubscriber & subscriber) {
+  std::cout << "subscribed!\n";
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED, sampleReceived);
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::UNSUBSCRIBED, waitForSubscription);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED);
+}
+
+void waitForUnsubscribe(iox::popo::UntypedSubscriber & subscriber) {
+  std::cout << "unsubscribed from publisher\n";
+  myCallbackReactal.attachEvent(subscriber, iox::popo::SubscriberEvent::SUBSCRIBED, waitForSubscription);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::HAS_SAMPLE_RECEIVED);
+  myCallbackReactal.detachEvent(subscriber, iox::popo::SubscriberEvent::UNSUBSCRIBED);
+}
+
+int main() {
+  myCallbackReactal.attachEvent(mySubscriber, iox::popo::SubscriberEvent::SUBSCRIBED, waitForSubscription);
+}
+
+```
+
+### Solution
+#### Condition Variable
+
+  - Create from `ConditionVariableWithOriginTracing` (yeah we need a better name) has `ConditionVariableData` 
```

Comment:
`EventSignaler` `EventWaiter`?

Reply:
Sounds better but I am still not perfectly happy.